sched/scheduler/task_scheduler.go

```diff
@@ -9,27 +9,50 @@ type taskAssignments struct {
 	task *taskState
 }
 
-// Returns a list of taskAssigments of task to available node.  Not all
-// tasks are guaranteed to be scheduled.  Does best effort scheduling
-func getTaskAssignments(cs *clusterState, tasks []*taskState) []taskAssignments {
-
-	// for now just assign any task to any available node
+// Returns a list of taskAssigments of task to available node.
+// Also returns a modified copy of clusterState.Affinity for the caller to apply (so this remains a pure fn).
+// Does best effort scheduling which tries to assign tasks to nodes already primed for similar tasks.
+// Not all tasks are guaranteed to be scheduled.
+func getTaskAssignments(cs *clusterState, tasks []*taskState) ([]taskAssignments, map[string]*nodeSet) {
 	var tas []taskAssignments
-	taskIndex := 0
+	affinity := map[string]*nodeSet{}
+	snapshotIds := []string{}
+
+	// Create a copy of affinity to modify based on new scheduling.
+	for snapId, set := range cs.affinity {
+		affinity[snapId] = newNodeSet()
+		for nodeId, node := range set.idle {
+			affinity[snapId].idle[nodeId] = node
```

Comment:
is there a race condition between this and

sched/scheduler/cluster_state.go#66
`delete(c.affinity[nodeState.snapshotID].idle, update.Id)`